storage/txcache/concurrentMap.go

```diff
@@ -0,0 +1,134 @@
+package txcache
+
+import (
+	"sync"
+)
+
+// ConcurrentMap is a "thread" safe map of type string:Anything.
+// To avoid lock bottlenecks this map is dived to several (ShardCount) map shards.
+type ConcurrentMap struct {
+	shardCount int
+	shards     []*ConcurrentMapShard
+}
+
+// ConcurrentMapShard is a "thread" safe string to anything map.
+type ConcurrentMapShard struct {
+	maxSize      int
+	items        map[string]interface{}
+	sync.RWMutex // Read Write mutex, guards access to internal map.
+}
+
+// NewConcurrentMap creates a new concurrent map.
+func NewConcurrentMap(maxSize int, shardCount int) *ConcurrentMap {
+	m := ConcurrentMap{
+		shardCount: shardCount,
+		shards:     make([]*ConcurrentMapShard, shardCount),
+	}
+
+	shardSize := maxSize / shardCount
+	if shardSize == 0 {
+		shardSize = 1
+	}
+	if maxSize%shardCount != 0 {
+		shardSize++
+	}
+	for i := 0; i < shardCount; i++ {
+		m.shards[i] = &ConcurrentMapShard{
+			maxSize: shardSize,
+
+			items: make(map[string]interface{}),
+		}
+	}
+	return &m
+}
+
+// GetShard returns shard under given key.
+func (m *ConcurrentMap) GetShard(key string) *ConcurrentMapShard {
+	return m.shards[uint(fnv32(key))%uint(m.shardCount)]
+}
+
+// Set sets the given value under the specified key.
+func (m *ConcurrentMap) Set(key string, value interface{}) {
+	// Get map shard.
+	shard := m.GetShard(key)
+	shard.Lock()
+	shard.items[key] = value
+	shard.Unlock()
+}
+
+// Get retrieves an element from map under given key.
+func (m *ConcurrentMap) Get(key string) (interface{}, bool) {
+	// Get shard
+	shard := m.GetShard(key)
+	shard.RLock()
+	// Get item from shard.
+	val, ok := shard.items[key]
```

Comment:
if !ok return nil

Reply:
`val` is already `<nil>` when key is missing (a bit less clutter with forwarding (returning) `val`, `ok` in my opinion).